### DISCUSSION TITLE
Does anonymization like replace/hash maintain a mapping that can help in deanonymization?

### DISCUSSION THREAD

**Discussion by himsgpt** [microsoft/presidio#1138] [Q&A]
I have created a pipeline as below; now how can I get the mapping used during anonymization that can help me deanonymize the data back to original form?
very little information around deanonymization is given!

1. {Input text }--> Analyzer(recognizers) ---> Anonymizer (using **replace, hashing**) --> {anonymized text}

2. {anonymized text} ---> original form (deanonymized)

**Comment by edwardcjohnson** [comment]
Please reference this example:
https://microsoft.github.io/presidio/samples/python/encrypt_decrypt/

**Reply by himsgpt** [reply]
I have seen this, two issues with this approach
- This is only valid for encryption technique and not for replace/hash/mask etc
- This approach will deanonymize the same text, while the requirement would be to deanonymize any text with encrypted information

ex on 2nd point: I encrypted the text and it should be able to work as below
Original text -  **My name is James bond**  as 
Encrypted text- **My name is M4lla0kBCzu6SwCONL6Y+ZqsPqhBp1Lhdc3t0FKnUwM=.**

Input for decryption -- "**M4lla0kBCzu6SwCONL6Y+ZqsPqhBp1Lhdc3t0FKnUwM=. enacted in the movies that featured ...**"
decrypted -- **James bond enacted in the movies that featured ..**

**Reply by edwardcjohnson** [reply]
i see what you mean. I think you'd have to write that functionality yourself by using a dictionary in which the key is the encrypted substring and the value is the unencrypted substring. Then you can easily find and replace upon decryption.

**Reply by himsgpt** [reply]
That could be one way. Although it would mean not using Presidio for deanonymization.
Thanks!

**Reply by omri374** [reply]
That's correct, the Deanonymize engine is meant to reverse the AnonymizerEngine's logic, and not to generalize to any text with encrypted substrings.

**Comment by omri374** [comment]
FYI, we have a new sample for pseudonymization:  https://github.com/microsoft/presidio/blob/main/docs/samples/python/pseudonomyzation.ipynb
